sdk/communication/azure-communication-common/src/main/java/com/azure/communication/common/implementation/RedirectPolicy.java

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.communication.common.implementation;
+
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * HttpPipelinePolicy to redirect requests when 302 message is received to the new location marked by the
+ * Location header.
+ */
+public final class RedirectPolicy implements HttpPipelinePolicy {
+    private static final int MAX_REDIRECTS = 10;
+    private static final String LOCATION_HEADER_NAME = "Location";
+    private final List<String> locations = new ArrayList<>();
+
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        return attemptRedirection(context, next, 0);
+    }
+
+    private Mono<HttpResponse> attemptRedirection(HttpPipelineCallContext context, HttpPipelineNextPolicy next,
+                                                  int redirectNumber) {
+        return next.clone().process().flatMap(httpResponse -> {
+            if (shouldRedirect(httpResponse, redirectNumber)) {
+                String newLocation = httpResponse.getHeaderValue(LOCATION_HEADER_NAME);
+                locations.add(newLocation);
+
+                HttpRequest newRequest = context.getHttpRequest().copy();
+                newRequest.setUrl(newLocation);
+                context.setHttpRequest(newRequest);
+
+                return attemptRedirection(context, next, redirectNumber + 1);
+            }
+            return Mono.just(httpResponse);
+        });
+    }
+
+    private boolean shouldRedirect(HttpResponse response, int redirectNumber) {
+        return response.getStatusCode() == 302
+            && !locations.contains(response.getHeaderValue(LOCATION_HEADER_NAME))
```

Comment:
Based on this a request isn't able to be redirected back to a previously used redirection? Is there a possibility that this causes an error?

Reply:
Assuming that the services are deterministic, if we allow to try a previously used location we will end up on some loop. So, we are short circuiting at the first opportunity.

Reply:
I think starting with that logic works, another possibility is making the previously used and failed redirects a chain, so if the service attempts to redirect you to somewhere that has been known to redirect can short-circuit the redirection chain isn't of failing the request outright, ex:

`location 1 -> 2 -> 3 -> 4 -> 5 -> success`

If the redirect is for `3` we could short-circuit to `5`. But this introduces a lot more complexity and chances of incorrect behavior.